File: CPP/Anagram.cpp
#include <iostream>
#include <algorithm>

using namespace std;

bool isAnagram(string str1, string str2){
    sort(str1.begin(), str1.end());
    sort(str2.begin(), str2.end());
    return str1 == str2;
}

string bubbleSortString(string str){
    for(int i = 0; i < str.length(); i++){
        for(int j = 0; j < str.length() - 1 - i; j++){
            if(str[j] > str[j+1]){
                swap(str[j], str[j+1]);
            }
        }
    }
    return str;
}

bool isAnagramBubbleSort(string str1, string str2){    
    return bubbleSortString(str1) == bubbleSortString(str2);
}

bool isAnagramCount(string str1, string str2){
    int length1 = str1.length();
    int length2 = str2.length();
    if(length1 != length2){
        return false;
    }
    int count[128] = {};
    for(int i = 0; i < length1; i++){
        count[str1[i]]++;
        count[str2[i]]--;
    }
    for(int i = 0; i < 128; i++){
        if(count[i]){
            return false;
        }
    }
    return true;
}

int main() {
    cout << isAnagram("listen", "silent") << endl;
	cout << isAnagram("angel", "glean") << endl;
	cout << isAnagram("brags", "grabs") << endl;
    
    cout << isAnagramBubbleSort("listen", "silent") << endl;
	cout << isAnagramBubbleSort("angel", "glean") << endl;
	cout << isAnagramBubbleSort("brags", "grabs") << endl;

    cout << isAnagramCount("listen", "silent") << endl;
	cout << isAnagramCount("angel", "glean") << endl;
	cout << isAnagramCount("brags", "grabs") << endl;
}

File: Python/Anagram.py
#O(nlogn)
def is_anagram(str1: str, str2: str) -> bool:
    return sorted(str1) == sorted(str2)

print(is_anagram("listen", "silent"))
print(is_anagram("angel", "glean"))
print(is_anagram("brags", "grabs"))

#O(n^2)
def bubble_sort_string(string: str) -> str:
    characters = list(string)
    for i in range(len(characters)):
        for j in range(len(characters) - 1 - i):
            if characters[j] > characters[j + 1]:
                characters[j], characters[j + 1] = characters[j + 1], characters[j]
    return ''.join(characters)

#O(n^2)
def is_anagram_bubble_sort(str1: str, str2: str) -> bool:
    return bubble_sort_string(str1) == bubble_sort_string(str2)

print(is_anagram_bubble_sort("listen", "silent"))
print(is_anagram_bubble_sort("angel", "glean"))
print(is_anagram_bubble_sort("brags", "grabs"))

#O(n)
def is_anagram_count(str1: str, str2: str) -> bool:
    char_count = [0]*128
    if len(str1) != len(str2):
        return False
    for i in range(len(str1)):
        char_count[ord(str1[i])]+=1
        char_count[ord(str2[i])]-=1
    for count in char_count:
        if count:
            return False
    return True

print(is_anagram_count("listen", "silent"))
print(is_anagram_count("angel", "glean"))
print(is_anagram_count("brags", "grabs"))